integration-tests/build.gradle.kts

```diff
@@ -21,7 +21,26 @@ subprojects {
         implementation(project(":integration-tests"))
     }
 
-    val integrationTest by tasks.register<Test>("integrationTest") {
+    /**
+     * Dokka's integration test task is not cacheable because the HTML outputs
+     * it produces when running the tests are used for showcasing resulting documentation.
+     *
+     * At the moment there are two problems that do not allow to make it cacheable:
+     * 1. The task's inputs are such that changes in Dokka's code do not invalidate the cache,
+     *    because it is run with the same version of Dokka ("DOKKA_VERSION") on the same
+     *    test project inputs.
+     * 2. The tests generate HTML output which is then used to showcase documentation.
+     *    The outputs are usually copied to a location from which it will be served.
+     *    However, if the test is cacheable, it produces no outputs, so no documentation
+     *    to showcase. It needs to be broken into two separate tasks: one cacheable for running
+     *    the tests and producing HTML output, and another non-cacheable for copying the output.
+     *
+     * @see [org.jetbrains.dokka.it.TestOutputCopier] for more details on showcasing documentation
+     */
+    @DisableCachingByDefault
+    abstract class NonCacheableIntegrationTest : Test()
+
+    val integrationTest by tasks.register<NonCacheableIntegrationTest>("integrationTest") {
         maxHeapSize = "2G"
         description = "Runs integration tests."
         group = "verification"
```

Comment:
```suggestion
        group = "verification"
        inputs.property("dokkaTestOutputPath", proiders.environmentVariable("DOKKA_TEST_OUTPUT_PATH"))
```

Reply:
This should help Gradle to determine if the test task is up-to-date and requires running again. It's not a complete solution though.

Reply:
I'm a bit confused on this, from @IgnatBeresnev's big comment in the code it sounds like this is an output dir, and an input param. Would adding
```
outputs.dir(providers.environmentVariable("DOKKA_TEST_OUTPUT_PATH"))
```
make the caching work? (i.e. even if it's from cache, it would put the right files in the right place)

Reply:
Note that "up-to-date"-ness is defined as "for a given input hash the output hash matches". Remember `outputs.upToDateWhen { true }`.

Depending on how much I misunderstand the context either the above, or
```
inputs.dir(providers.environmentVariable("DOKKA_TEST_OUTPUT_PATH"))
```
might be also necessary, so that the contents are also considered, not just where it is.

Reply:
Anyway, sorry for the drive-by, I realize I don't have enough context on Dokka workflow to help here.

Reply:
I explained a little more here: https://github.com/Kotlin/dokka/pull/2826#discussion_r1089959966

Setting the value of `DOKKA_TEST_OUTPUT_PATH` might help, but the task would still need to figure out when the Dokka source code changes. Because the Dokka source code is published to Maven Local, Gradle can't figure out when the source code has changed.

I suspect a more complete solution would involve

* setting up a 'Maven Publication Test' Gradle convention plugin, that will 
   1. publish the Dokka project to a project-local directory instead of Maven Local
   2. use Gradle Configurations to share the result of publications
* and using the JVM Test Suite plugin, or using `includeBuild()`, to better define and manage the integration tests

Which is complicated, and needs some prep-work first #2703, so yeah, for now it's best just to tell Gradle not to cache the integration tests.

Reply:
Might be a silly thing, but using the `-sources.jar` as input?

Reply:
Possibly...? I'm not sure, I've not shared sources.jars before. I think that might also require some custom Gradle Configurations? And then you might as well just have Configurations with the Maven Publication, which would be more representative.

Reply:
> This should help Gradle to determine if the test task is up-to-date and requires running again. It's not a complete solution though.

I'm not sure this would help - doesn't the value of the property have to change for the task to be invalidated/cached? There's only 3 possible values for `DOKKA_TEST_OUTPUT_PATH`, you can see them [here](https://github.com/Kotlin/dokka/blob/master/.github/workflows/gh-actions-artifacts-snapshots.yml#L28). I would expect that Gradle would have 3 different cache entries for all 3 values, it doesn't just store the last run, right? 

And `DOKKA_TEST_OUTPUT_PATH` is GitHub action specific, so it wouldn't help with the problem for local runs and for TeamCity runs if it starts using Gradle caches.

Anyhow, I think it's best to disable the test for now as it's not that critical, and spend time re-writing this part completely instead of trying to patch it up. Just need to get all of the other Gradle-related things in order first :sweat_smile: